### PR TITLE
Added new `detect-child-process` rule (#252)

### LINKED ISSUE #252: new security rule: detect-child-process
https://github.com/nodesecurity/eslint-plugin-security/blob/master/rules/detect-child-process.js

### PR DESCRIPTION
**Pull request by soon**
#### PR checklist

-   [x] Addresses an existing issue: fixes #252
-   [x] New feature, bugfix, or enhancement
    -   [x] Includes tests
-   [x] Documentation update

#### Overview of change:

Added new `detect-child-process` rule.

#### Is there anything you'd like reviewers to focus on?

Spelling (docs), Rule meta (not sure about issueClass, severity and level).

### CODE REVIEW

File: src/detectChildProcessRule.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import * as ts from 'typescript';
+import { AstUtils } from './utils/AstUtils';
+
+import { ExtendedMetadata } from './utils/ExtendedMetadata';
+
+const FORBIDDEN_IMPORT_FAILURE_STRING: string = 'Found child_process import';
+const FOUND_EXEC_FAILURE_STRING: string = 'Found child_process.exec() with non-literal first argument';
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static metadata: ExtendedMetadata = {
```
Comment (JoshuaKGoldberg) [inline src/detectChildProcessRule.ts]: This should also have a `rationale`, which will be particularly helpful once rule documentation is auto-generated. You can generally copy & paste the rough description from the README.md table.

File: src/detectChildProcessRule.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        ruleName: 'detect-child-process',
+        type: 'maintainability',
+        description: 'Tries to detect instances of child_process',
```
Comment (JoshuaKGoldberg) [inline src/detectChildProcessRule.ts]:
```suggestion
        description: 'Detects instances of child_process and child_process.exec',
```

File: src/detectChildProcessRule.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        }
+        importedFunctionsAliases.forEach(x => childProcessFunctionAliases.add(x));
+    }
+}
+
+function walk(ctx: Lint.WalkContext<void>) {
+    const childProcessModuleAliases = new Set<string>();
+    const childProcessFunctionAliases = new Set<string>();
+
+    function cb(node: ts.Node): void {
```
Comment (JoshuaKGoldberg) [inline src/detectChildProcessRule.ts]: This `cb` function is pretty large and getting a little hard to read. Could you please split the major parts of it into separate functions? They can be declared within the parent `walk` function if you'd like.

Example: `if (tsutils.isImportEqualsDeclaration(node)) {` declares variables within itself, so it's a good candidate for splitting out.

File: tests/detect-child-process/test.ts.lint [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Found child_process import]
+import {spawn} from 'child_process';
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Found child_process import]
+
+import {exec as someAnotherExec} from 'child_process';
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Found child_process import]
+import {spawn as someAnotherSpawn} from 'child_process';
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Found child_process import]
+
+import * as anotherModule from 'anotherModule';
```
Comment (JoshuaKGoldberg) [inline tests/detect-child-process/test.ts.lint]: Missing a test cases for `require`s.

```ts
// Examples
const child_process = require("child_process");
const { exec } = require("child_process");
const { spawn } = require("child_process");
```

File: src/detectChildProcessRule.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        group: 'Security',
+        commonWeaknessEnumeration: '88'
+    };
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk);
+    }
+}
+
+function processImport(
```
Comment (JoshuaKGoldberg) [inline src/detectChildProcessRule.ts]: Suggestion: move this into the `walk` function so it has access to `childProcessModuleAliases` and `childProcessFunctionAliases`.

File: tests/detect-child-process/test.ts.lint [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+
+child_process.exec('ls')
+child_process.exec(cmd)
```
Comment (JoshuaKGoldberg) [inline tests/detect-child-process/test.ts.lint]: Missing test case: what if additional options are passed in?

The source code checks if `node.arguments.length === 1`, which should probably be changed to `!== 0`.

**Comment by JoshuaKGoldberg** [comment]
Sorry for taking so long to get to this! I'm happy to discuss here or on gitter.im if any of the feedback doesn't make sense!

File: README.md [hunk 7]
```diff
@@ -131,6 +131,18 @@ We recommend you specify exact versions of lint libraries, including `tslint-mic
...
     </tr>
+    <tr>
+      <td>
+        <code>detect-child-process</code>
+      </td>
+      <td>
+        Detects usages of child_process and especially child_process.exec() with a non-literal first argument.
+        It is dangerous to pass a string constructed at runtime as the first argument to the child_process.exec().
+        `child_process.exec(cmd)` runs `cmd` as a shell command which allows attacker to execute malicious code injected into `cmd` string.
+        Instead of `child_process.exec(cmd)` you should use `child_process.spawn(cmd)` or specify the command as a literal, e.g. `child_process.exec('ls')`.
```
[the remark this answers is not in the document]
Comment (soon) [reply README.md]: Fixed

File: README.md [hunk 8]
```diff
@@ -131,6 +131,18 @@ We recommend you specify exact versions of lint libraries, including `tslint-mic
...
       <td>1.0</td>
     </tr>
+    <tr>
+      <td>
+        <code>detect-child-process</code>
+      </td>
+      <td>
+        Detects usages of child_process and especially child_process.exec() with a non-literal first argument.
+        It is dangerous to pass a string constructed at runtime as the first argument to the child_process.exec().
+        `child_process.exec(cmd)` runs `cmd` as a shell command which allows attacker to execute malicious code injected into `cmd` string.
```
[the remark this answers is not in the document]
Comment (soon) [reply README.md]: Fixed

File: src/detectChildProcessRule.ts (hunk 1 above)
Comment (soon) [reply src/detectChildProcessRule.ts]: Fixed

File: src/detectChildProcessRule.ts (hunk 2 above)
Comment (soon) [reply src/detectChildProcessRule.ts]: Fixed

File: src/detectChildProcessRule.ts (hunk 3 above)
Comment (soon) [reply src/detectChildProcessRule.ts]: Done

File: tests/detect-child-process/test.ts.lint (hunk 4 above)
Comment (soon) [reply tests/detect-child-process/test.ts.lint]: Done

File: src/detectChildProcessRule.ts (hunk 5 above)
Comment (soon) [reply src/detectChildProcessRule.ts]: Done

File: tests/detect-child-process/test.ts.lint (hunk 6 above)
Comment (soon) [reply tests/detect-child-process/test.ts.lint]: Good catch, fixed

**Comment by soon** [comment]
@JoshuaKGoldberg Thanks for reviewing! I've just updated the code, please review

Do we need to handle `require(['child_process'])`? I never used this form of `require`, but I found some tests including require with array as the first argument for other checks

**Comment by JoshuaKGoldberg** [comment]
> Do we need to handle require(['child_process'])? I've never used this form of require before, but I found some tests including require with array as the first argument for other checks

Ooh, good question. No, let's ignore that for now. Direct AMD modules are falling out of favor and most rules don't bother with them.

**Comment by soon** [comment]
@JoshuaKGoldberg Anything else should be changed in this PR?

File: README.md [hunk 9]
```diff
@@ -131,6 +131,18 @@ We recommend you specify exact versions of lint libraries, including `tslint-mic
       </td>
       <td>1.0</td>
     </tr>
+    <tr>
+      <td>
+        <code>detect-child-process</code>
+      </td>
+      <td>
+        Detects usages of child_process and especially child_process.exec() with a non-literal first argument.
+        It is dangerous to pass a string constructed at runtime as the first argument to the child_process.exec().
```
[the remark this answers is not in the document]
Comment (soon) [reply README.md]: Fixed

File: README.md [hunk 10]
```diff
@@ -131,6 +131,18 @@ We recommend you specify exact versions of lint libraries, including `tslint-mic
...
+        <code>child_process.exec(cmd)</code> runs <code>cmd</code> as a shell command which allows attacker to execute malicious code injected into <code>cmd</code> string.
```
[the remark this answers is not in the document]
Comment (soon) [reply README.md]: Fixed

**Comment by soon** [comment]
@JoshuaKGoldberg Commited your changes

**Pull request merged**